packages/bits/src/lib/sorter/sorter-key-control.service.ts

```diff
@@ -0,0 +1,100 @@
+import { Injectable, QueryList } from "@angular/core";
+import { ListKeyManager, LiveAnnouncer } from "@angular/cdk/a11y";
+import isNull from "lodash/isNull";
+
+import { MenuItemBaseComponent } from "../menu";
+import { OverlayComponent } from "../overlay/overlay-component/overlay.component";
+import { KEYBOARD_CODE } from "../../constants";
+import { ANNOUNCER_CLOSE_SORTER_LIST_MESSAGE, ANNOUNCER_OPEN_SORTER_LIST_MESSAGE_SUFFIX } from "./constants";
+
+@Injectable()
+export class SorterKeyControlService {
+    public overlay: OverlayComponent;
+    public menuItems: QueryList<MenuItemBaseComponent>;
+    private keyboardEventsManager: ListKeyManager<MenuItemBaseComponent>;
+
+    constructor(private liveAnnouncer: LiveAnnouncer) {}
+
+    public initKeyboardManager(): void {
+        this.keyboardEventsManager = new ListKeyManager(this.menuItems).withVerticalOrientation();
+        // TODO Uncomment in the scope of NUI-
+        // this.keyboardEventsManager.setFirstItemActive();
+
+        // TODO Remove this in the scope of NUI-
+        this.keyboardEventsManager.setActiveItem(-1);
+    }
+
+    public handleKeydown(event: KeyboardEvent): void {
+        this.overlay.showing ? this.handleOpenKeyDown(event) : this.handleClosedKeyDown(event);
+    }
+
+    public getActiveItemIndex(): number | null {
+        return this.keyboardEventsManager.activeItemIndex;
+    }
+
+    public announceDropdown(open: boolean): void {
+        const menuListLength = this.menuItems.toArray().length;
+        const message = open ? `${menuListLength} ${ANNOUNCER_OPEN_SORTER_LIST_MESSAGE_SUFFIX}` : ANNOUNCER_CLOSE_SORTER_LIST_MESSAGE;
+        this.liveAnnouncer.announce(message);
+    }
+
+    private handleOpenKeyDown(event: KeyboardEvent): void {
+        this.keyboardEventsManager.activeItem?.setInactiveStyles();
+        if (event.code === KEYBOARD_CODE.ARROW_DOWN || event.code === KEYBOARD_CODE.ARROW_UP) {
+            this.keyboardEventsManager.onKeydown(event);
+            this.announceCurrentItem();
+        }
+
+        // TODO Remove this in the scope of NUI-
+        // prevent closing on enter when item is not focused
+        if (!this.hasActiveItem() && event.code === KEYBOARD_CODE.ENTER) {
+            event.preventDefault();
+        }
+
+        if (this.hasActiveItem() && event.code === KEYBOARD_CODE.ENTER) {
+            event.preventDefault();
+            if (!this.keyboardEventsManager.activeItem) {
+                throw new Error("ActiveItem is not defined");
```

Comment:
I think we can avoid this check because we're already assured that we have an active item thanks to the `hasActiveItem() `inside the `if` condition. What do you think?